runtime/gc_vlhgc/EnvironmentVLHGC.hpp

```diff
@@ -100,6 +99,84 @@ class MM_EnvironmentVLHGC : public MM_EnvironmentBase
 	MMINLINE static MM_EnvironmentVLHGC *getEnvironment(OMR_VMThread *omrVMThread) { return static_cast<MM_EnvironmentVLHGC*>(omrVMThread->_gcOmrVMThreadExtensions); }
 	MMINLINE static MM_EnvironmentVLHGC *getEnvironment(MM_EnvironmentBase* env) { return static_cast<MM_EnvironmentVLHGC*>(env); }
 
+	MMINLINE static uintptr_t rsCardSize(bool compressed) { return compressed ? sizeof(uint32_t) : sizeof(uintptr_t); }
+
+	MMINLINE uintptr_t rsCardSize() { return rsCardSize(compressObjectReferences()); }
+
+	/**
+	 * Read the value of a card.
+	 *
+	 * @param[in] addr the card address
+	 * @return the card value
+	 */
+	MMINLINE MM_RememberedSetCard readCard(MM_RememberedSetCard *addr)
+	{
+		return compressObjectReferences() ? (MM_RememberedSetCard)*(uint32_t*)addr : (MM_RememberedSetCard)*(uintptr_t*)addr;
+	}
+
+	/**
+	 * Write the value of a card.
+	 *
+	 * @param[in] addr the card address
+	 * @param[in] card the card value
+	 */
+	MMINLINE void writeCard(MM_RememberedSetCard *addr, MM_RememberedSetCard card)
+	{
+		if (compressObjectReferences()) {
+			*(uint32_t*)addr = (uint32_t)card;
+		} else {
+			*(uintptr_t*)addr = (uintptr_t)card;
+		}
+	}
+
+	/**
+	 * Calculate the addition of an integer to a card address
+	 *
+	 * @param[in] base the base card address
+	 * @param[in] index the index to add
+	 * @return the adjusted address
+	 */
+	MMINLINE MM_RememberedSetCard* addToCardAddress(MM_RememberedSetCard *base, intptr_t index)
+	{
+		if (compressObjectReferences()) {
+			return (MM_RememberedSetCard*)((uint32_t*)base + index);
+		} else {
+			return (MM_RememberedSetCard*)((uintptr_t*)base + index);
+		}
+	}
+
+	/**
+	 * Calculate the subtraction of an integer from a card address
+	 *
+	 * @param[in] base the base card address
+	 * @param[in] index the index to subtract
+	 * @return the adjusted address
+	 */
+	MMINLINE MM_RememberedSetCard* subtractFromCardAddress(MM_RememberedSetCard *base, intptr_t index)
+	{
+		if (compressObjectReferences()) {
+			return (MM_RememberedSetCard*)((uint32_t*)base - index);
+		} else {
+			return (MM_RememberedSetCard*)((uintptr_t*)base - index);
+		}
+	}
+
+	/**
+	 * Calculate the difference between two card addresses, in slots
+	 *
+	 * @param[in] p1 the value to be subtracted from
+	 * @param[in] p2 the value to be subtracted
+	 * @return p1 - p2 in slots
+	 */
+	MMINLINE intptr_t subtractCardAddresses(MM_RememberedSetCard *p1, MM_RememberedSetCard *p2)
+	{
+		if (compressObjectReferences()) {
+			return (uint32_t*)p1 - (uint32_t*)p2;
+		} else {
+			return (uintptr_t*)p1 - (uintptr_t*)p2;
+		}
+	}
+
```

Comment:
I might have had been ok with this, but thinking now if these 4 methods (read/write/sub/add) could be statics in RSCard (which obviously needs to upgrade from typedef to class)?

That would be at least consistent with GC_SlotObject story.

Reply:
Let's discuss this in person - it's not as simple as making the typedef into a class (we won't have instances of the class).